dot_emacs.d/configuration.org
#+TITLE: Emacs Configuration
#+AUTHOR: Zach
#+EMAIL:
#+OPTIONS: toc:nil num:nil
This is a test org config file that's a frankenstein from
 https://github.com/hrs/dotfiles/blob/main/emacs/dot-emacs.d/configuration.org
which is much more organized and thought out.

* Configure =use-package=

I use =use-package= to install and configure my packages. My =init.el= includes
the initial setup for =package.el= and ensures that =use-package= is installed,
since I wanna do that right away.

This makes sure that =use-package= will install the package if it's not already
available. It also means that I should be able to open Emacs for the first time
on a fresh Debian box and have my whole environment automatically installed. I'm
not /totally/ sure about that, but we're gettin' close.

#+begin_src emacs-lisp
  (require 'use-package-ensure)
  (setq use-package-always-ensure t)
#+end_src

Always compile packages, and use the newest version available.

#+begin_src emacs-lisp
  (use-package auto-compile
    :config (auto-compile-on-load-mode))
  (setq load-prefer-newer t)
#+end_src

* Set personal information

** Who am I? Where am I?

#+begin_src emacs-lisp
  (setq user-full-name "Zachary Bombay"
        user-mail-address "zacharybombay@gmail.com"
        calendar-latitude 0.0
        calendar-longitude 0.0
        calendar-location-name "")
#+end_src

# * Add =resources= to =load-path=

# #+begin_src emacs-lisp
#   (add-to-list 'load-path "~/.emacs.d/resources/")
# #+end_src

* Utility functions

Define a big ol' bunch of handy utility functions.

#+begin_src emacs-lisp
  (defun zb/rename-file (new-name)
    (interactive "FNew name: ")
    (let ((filename (buffer-file-name)))
      (if filename
    (progn
      (when (buffer-modified-p)
        (save-buffer))
      (rename-file filename new-name t)
      (kill-buffer (current-buffer))
      (find-file new-name)
      (message "Renamed '%s' -> '%s'" filename new-name))
  (message "Buffer '%s' isn't backed by a file!" (buffer-name)))))

  (defun zb/append-to-path (path)
    "Add a path both to the $PATH variable and to Emacs' exec-path."
    (setenv "PATH" (concat (getenv "PATH") ":" path))
    (add-to-list 'exec-path path))

  (defun zb/notify-send (title message)
    "Display a desktop notification by shelling out to `notify-send'."
    (call-process-shell-command
     (format "notify-send -t 2000 \"%s\" \"%s\"" title message)))
#+end_src

* UI preferences

** UTF-8 complience

#+begin_src emacs-lisp
  (prefer-coding-system 'utf-8)
  (set-default-coding-systems 'utf-8)
  (set-terminal-coding-system 'utf-8)
  (setq-default buffer-file-coding-system 'utf-8)
#+end_src

** Tweak window chrome

I don't usually use the menu or scroll bar, and they take up useful space.

#+begin_src emacs-lisp
  (tool-bar-mode 0)
  (menu-bar-mode 0)
  (scroll-bar-mode -1)
  ;  (global-linum-mode t)
  (column-number-mode t)
  (setq inhibit-startup-message t)
  (global-visual-line-mode)
  (setq frame-title-format "%b")
;;  (setq frame-title-format '((:eval (projectile-project-name))))
  (add-hook 'before-save-hook 'delete-trailing-whitespace)
#+end_src

** Ido mode

#+begin_src emacs-lisp
  (setq ido-enable-flex-matching t)
  (setq ido-everywhere t)
  (ido-mode 1)
#+end_src

** Use fancy lambdas

Why not?

#+begin_src emacs-lisp
  (global-prettify-symbols-mode t)
#+end_src

** Highlight the current line

=global-hl-line-mode= softly highlights the background color of the line
containing point. It makes it a bit easier to find point, and it's useful when
pairing or presenting code.

#+begin_src emacs-lisp
  (global-hl-line-mode)
#+end_src

** Load up a theme
#+begin_src emacs-lisp
  (use-package tron-legacy-theme)
  (load-theme 'tron-legacy t)
#+end_src

    # abyss-theme
    # underwater-theme
    # challenger-deep-theme
    # cyberpunk-theme
    # dakrone-theme
    # dracula-theme
    # espresso-theme
    # exotica-theme

** Use =moody= for a beautiful modeline

This gives me a truly lovely ribbon-based modeline.

#+begin_src emacs-lisp
    (use-package moody
      :config
      (setq x-underline-at-descent-line t
	    moody-mode-line-height 30)
    (moody-replace-mode-line-buffer-identification))
;;    (moody-replace-vc-mode))
#+end_src

** Use =minions= to hide all minor modes

I never want to see a minor mode, and manually adding =:diminish= to every
use-package declaration is a hassle. This uses =minions= to hide all the minor
modes in the modeline. Nice!

By default there's a =;-)= after the major mode; that's an adorable default, but
I'd rather skip it.

#+begin_src emacs-lisp
   (use-package minions
     :config
     (setq minions-mode-line-lighter ""
           minions-mode-line-delimiters '("" . ""))
     (minions-mode 1))
#+end_src

** Set default font and configure font resizing

I'm partial to Source Code Pro.

The standard =text-scale-= functions just resize the text in the current buffer;
I'd generally like to resize the text in /every/ buffer, and I usually want to
change the size of the modeline, too (this is especially helpful when
presenting). These functions and bindings let me resize everything all together!

Note that this overrides the default font-related keybindings from
=sensible-defaults=.

#+begin_src emacs-lisp
  (setq zb/default-font "Source Code Pro")
  (setq zb/default-font-size 12)
  (setq zb/current-font-size zb/default-font-size)

  (setq zb/font-change-increment 1.1)

  (defun zb/font-code ()
    "Return a string representing the current font (like \"Source Code Pro-14\")."
    (concat zb/default-font "-" (number-to-string zb/current-font-size)))

  (defun zb/set-font-size ()
    "Set the font to `zb/default-font' at `zb/current-font-size'.
  Set that for the current frame, and also make it the default for
  other, future frames."
    (let ((font-code (zb/font-code)))
      (if (assoc 'font default-frame-alist)
	  (setcdr (assoc 'font default-frame-alist) font-code)
	(add-to-list 'default-frame-alist (cons 'font font-code)))
      (set-frame-font font-code)))

  (defun zb/reset-font-size ()
    "Change font size back to `zb/default-font-size'."
    (interactive)
    (setq zb/current-font-size zb/default-font-size)
    (zb/set-font-size))

  (defun zb/increase-font-size ()
    "Increase current font size by a factor of `zb/font-change-increment'."
    (interactive)
    (setq zb/current-font-size
	  (ceiling (* zb/current-font-size zb/font-change-increment)))
    (zb/set-font-size))

  (defun zb/decrease-font-size ()
    "Decrease current font size by a factor of `zb/font-change-increment', down to a minimum size of 1."
    (interactive)
    (setq zb/current-font-size
	  (max 1
	  (floor (/ zb/current-font-size zb/font-change-increment))))
    (zb/set-font-size))

  (define-key global-map (kbd "C-)") 'zb/reset-font-size)
  (define-key global-map (kbd "C-+") 'zb/increase-font-size)
  (define-key global-map (kbd "C--") 'zb/decrease-font-size)
  (zb/reset-font-size)
#+end_src


** Highlight uncommitted changes

Use the =diff-hl= package to highlight changed-and-uncommitted lines when
programming.

#+begin_src emacs-lisp
  (use-package diff-hl
    :config
    (add-hook 'prog-mode-hook 'turn-on-diff-hl-mode)
    (add-hook 'vc-dir-mode-hook 'turn-on-diff-hl-mode))
#+end_src

* Programming Project Management

I use a few packages in virtually every programming or writing environment to
manage the project, handle auto-completion, search for terms, and deal with
version control. That's all in here.

** =company=

Use =company-mode= everywhere.

#+begin_src emacs-lisp
  (use-package company)
  (add-hook 'after-init-hook 'global-company-mode)
#+end_src

Use =M-/= for completion.

#+begin_src emacs-lisp
  (global-set-key (kbd "M-/") 'company-complete-common)
#+end_src

** =flycheck=

I'd like to enable flycheck all kinds of places.

#+begin_src emacs-lisp
  (use-package let-alist)
  (use-package flycheck
    :hook ruby-mode)
#+end_src

** =magit=

I use =magit= to handle version control. It's lovely, but I tweak a few things:

- I bring up the status menu with =C-x g=.
- Use =evil= keybindings with =magit=.
- The default behavior of =magit= is to ask before pushing. I haven't had any
  problems with accidentally pushing, so I'd rather not confirm that every time.
- Per [[http://tbaggery.com/2008/04/19/a-note-about-git-commit-messages.html][tpope's suggestions]], highlight commit text in the summary line that goes
  beyond 50 characters.
- I'd like to start in the insert state when writing a commit message.

#+begin_src emacs-lisp
  (use-package magit
    :bind
    ("C-x g" . magit-status)

    :config
    (use-package with-editor)

    (setq magit-push-always-verify nil
	  git-commit-summary-max-length 50))
#+end_src


* Programming environments

I like shallow indentation, but tabs are displayed as 8 characters by default.
This reduces that.

#+begin_src emacs-lisp
  (setq-default tab-width 2)
#+end_src

Treating terms in CamelCase symbols as separate words makes editing a little
easier for me, so I like to use =subword-mode= everywhere.

#+begin_src emacs-lisp
  (use-package subword
    :config (global-subword-mode 1))
#+end_src

Compilation output goes to the =*compilation*= buffer. I rarely have that window
selected, so the compilation output disappears past the bottom of the window.
This automatically scrolls the compilation window so I can always see the
output.

#+begin_src emacs-lisp
  (setq compilation-scroll-output t)
#+end_src


** Golang

Install =go-mode= and related packages:

#+begin_src emacs-lisp
  (use-package go-mode)
  (use-package go-errcheck)
  (use-package company-go)
#+end_src

Define my =$GOPATH= and tell Emacs where to find the Go binaries.

#+begin_src emacs-lisp
  (setenv "GOPATH" "~/code/go")
  ;; (setenv "GOPATH" "/home/zach/code/go")
  (zb/append-to-path (concat (getenv "GOPATH") "/bin"))
#+end_src

Run =goimports= on every file when saving, which formats the file and
automatically updates the list of imports. This requires that the =goimports=
binary be installed.

#+begin_src emacs-lisp
  (setq gofmt-command "goimports")
  (add-hook 'before-save-hook 'gofmt-before-save)
#+end_src

When I open a Go file,

- Start up =company-mode= with the Go backend. This requires that the =gocode=
  binary is installed,
- Redefine the default =compile= command to something Go-specific, and
- Enable =flycheck=.

#+begin_src emacs-lisp
  (add-hook 'go-mode-hook
            (lambda ()
              (set (make-local-variable 'company-backends)
                   '(company-go))
              (company-mode)
              (if (not (string-match "go" compile-command))
                  (set (make-local-variable 'compile-command)
                       "go build -v && go test -v && go vet"))
              (flycheck-mode)))
#+end_src

** Haskell

# #+begin_src emacs-lisp
#   (use-package haskell-mode)
# #+end_src

# Enable =haskell-doc-mode=, which displays the type signature of a function, and
# use smart indentation.

# #+begin_src emacs-lisp
#   (add-hook 'haskell-mode-hook
#             (lambda ()
#               (haskell-doc-mode)
#               (turn-on-haskell-indent)))
# #+end_src

# #+begin_src emacs-lisp
#   (zb/append-to-path "~/.cabal/bin")
# #+end_src

** JavaScript and CoffeeScript

Install =coffee-mode= from editing CoffeeScript code.

# #+begin_src emacs-lisp
#   (use-package coffee-mode)
# #+end_src

# Indent everything by 2 spaces.

# #+begin_src emacs-lisp
#   (setq js-indent-level 2)

#   (add-hook 'coffee-mode-hook
#             (lambda ()
#               (yas-minor-mode 1)
#               (setq coffee-tab-width 2)))
# #+end_src

** Lisps

*** All Lisps

I like to use =paredit= in Lisp modes to balance parentheses (and more!).

#+begin_src emacs-lisp
  (use-package paredit)
#+end_src

=rainbow-delimiters= is convenient for coloring matching parentheses.

#+begin_src emacs-lisp
  (use-package rainbow-delimiters)
#+end_src

All the lisps have some shared features, so we want to do the same things for
all of them. That includes using =paredit=, =rainbow-delimiters=, and
highlighting the whole expression when point is on a parenthesis.

#+begin_src emacs-lisp
  (setq lispy-mode-hooks
        '(clojure-mode-hook
          emacs-lisp-mode-hook
          lisp-mode-hook
          scheme-mode-hook))

  (dolist (hook lispy-mode-hooks)
    (add-hook hook (lambda ()
                     (setq show-paren-style 'expression)
                     (paredit-mode)
                     (rainbow-delimiters-mode))))
#+end_src

If I'm writing in Emacs lisp I'd like to use =eldoc-mode= to display
documentation.

#+begin_src emacs-lisp
  (use-package eldoc
    :config
    (add-hook 'emacs-lisp-mode-hook 'eldoc-mode))
#+end_src

I also like using =flycheck-package= to ensure that my Elisp packages are
correctly formatted.

#+begin_src emacs-lisp
  (use-package flycheck-package)

  (eval-after-load 'flycheck
    '(flycheck-package-setup))
#+end_src

# ** OCaml

# Use =tuareg-mode= for editing OCaml.

# #+begin_src emacs-lisp
#   (use-package tuareg
#     :config
#     (electric-indent-mode 0))
# #+end_src

# Configure Merlin. This also requires installing the Merlin package through OPAM
# with =opam install merlin=.

# #+begin_src emacs-lisp
#   (use-package merlin
#     :config
#     (add-hook 'tuareg-mode-hook 'merlin-mode t)

#     (with-eval-after-load 'company
#       (add-to-list 'company-backends 'merlin-company-backend))
#     (add-hook 'merlin-mode-hook 'company-mode))
# #+end_src

** Python

Use python-mode package and set python executable path depending on OS.
#+begin_src emacs-lisp
  (use-package python-mode)
  (cond
    ((string-equal system-type "windows-nt")
      (progn
        (zb/append-to-path "C:\\Python38\\Scripts")))
    ((string-equal system-type "gnu/linux")
      (progn
        (zb/append-to-path "/bin"))))
;;        (zb/append-to-path "~/.local/bin"))))
#+end_src

Enable =elpy=. This provides automatic indentation, auto-completion, syntax
checking, etc.

#+begin_src emacs-lisp
  (use-package elpy)
  (elpy-enable)
  (setq python-shell-interpreter "python"
        python-shell-interpreter-args "-i")
#+end_src

Use =flycheck= for syntax checking:

#+begin_src emacs-lisp
  (add-hook 'elpy-mode-hook 'flycheck-mode)
#+end_src

Configure Jedi along with the associated =company= mode:

#+begin_src emacs-lisp
  (use-package company-jedi)
  (add-to-list 'company-backends 'company-jedi)

  (add-hook 'python-mode-hook 'jedi:setup)
  (setq jedi:complete-on-dot t)
#+end_src

** Rust

Use =rust-mode= to edit Rust code.

Run =rustfmt= automatically when saving a file.

#+begin_src emacs-lisp
  (use-package rust-mode
    :config
    (zb/append-to-path "~/.cargo/bin")
    (setq rust-format-on-save t))
#+end_src

** =sh=

Indent with 2 spaces.

#+begin_src emacs-lisp
  (add-hook 'sh-mode-hook
            (lambda ()
              (setq sh-basic-offset 2
                    sh-indentation 2)))
#+end_src

** Scala

# Ensure that =scala-mode= and =sbt-mode= are installed.

# #+begin_src emacs-lisp
#   (use-package scala-mode
#     :interpreter
#     ("scala" . scala-mode))
#   (use-package sbt-mode)
# #+end_src

# Don't show the startup message with launching ENSIME:

# #+begin_src emacs-lisp
#   (setq ensime-startup-notification nil)
# #+end_src

# Bind a few keys to common operations:

# #+begin_src emacs-lisp
#   (evil-define-key 'normal ensime-mode-map (kbd "C-t") 'ensime-type-at-point)
#   (evil-define-key 'normal ensime-mode-map (kbd "M-.") 'ensime-edit-definition)
# #+end_src


** YAML

#+begin_src emacs-lisp
  (use-package yaml-mode)
#+end_src

* Terminal

I add a bunch of hooks to =term-mode=:

- I'd like links (URLs, etc) to be clickable.
- Yanking in =term-mode= doesn't quite work. The text from the paste appears in
  the buffer but isn't sent to the shell process. This correctly binds =C-y= and
  middle-click to yank the way we'd expect.
- I bind =M-o= to quickly change windows. I'd like that in terminals, too.
- I don't want to perform =yasnippet= expansion when tab-completing.

#+begin_src emacs-lisp
  (defun zb/term-paste (&optional string)
    (interactive)
    (process-send-string
     (get-buffer-process (current-buffer))
     (if string string (current-kill 0))))

  (add-hook 'term-mode-hook
            (lambda ()
              (goto-address-mode)
              (define-key term-raw-map (kbd "C-y") 'zb/term-paste)
              (define-key term-raw-map (kbd "<mouse-2>") 'zb/term-paste)
              (define-key term-raw-map (kbd "M-o") 'other-window))
              (setq yas-dont-activate t))
#+end_src

* Publishing and task management with Org-mode

I'd like the initial scratch buffer to be in Org:

#+begin_src emacs-lisp
  (setq initial-major-mode 'org-mode)
#+end_src
** Org-Roam

#+begin_src emacs-lisp
  (use-package org-roam)
  (executable-find "sqlite3")
  (setq org-roam-directory "~/sync/Nextcloud/notes/org-roam")
  (add-hook 'after-init-hook 'org-roam-mode)
#+end_src

** Display preferences

I like to see an outline of pretty bullets instead of a list of asterisks.


dot_emacs.d/configuration.el
(require 'use-package-ensure)
(setq use-package-always-ensure t)

(use-package auto-compile
  :config (auto-compile-on-load-mode))
(setq load-prefer-newer t)

(setq user-full-name "Zachary Bombay"
      user-mail-address "zacharybombay@gmail.com"
      calendar-latitude 39.034655
      calendar-longitude -77.380606
      calendar-location-name "Sterling, VA")

(defun zb/rename-file (new-name)
  (interactive "FNew name: ")
  (let ((filename (buffer-file-name)))
    (if filename
	(progn
	  (when (buffer-modified-p)
	    (save-buffer))
	  (rename-file filename new-name t)
	  (kill-buffer (current-buffer))
	  (find-file new-name)
	  (message "Renamed '%s' -> '%s'" filename new-name))
(message "Buffer '%s' isn't backed by a file!" (buffer-name)))))

(defun zb/generate-scratch-buffer ()
  "Create and switch to a temporary scratch buffer with a random
 name."
  (interactive)
  (switch-to-buffer (make-temp-name "scratch-")))

(defun zb/find-file-as-sudo ()
  (interactive)
  (let ((file-name (buffer-file-name)))
    (when file-name
(find-alternate-file (concat "/sudo::" file-name)))))

(defun zb/region-or-word ()
  (if mark-active
(buffer-substring-no-properties (region-beginning)
				(region-end))
    (thing-at-point 'word)))

(defun zb/append-to-path (path)
  "Add a path both to the $PATH variable and to Emacs' exec-path."
  (setenv "PATH" (concat (getenv "PATH") ":" path))
  (add-to-list 'exec-path path))

(defun zb/insert-password ()
  (interactive)
  (shell-command "pwgen 30 -1" t))

(defun zb/notify-send (title message)
  "Display a desktop notification by shelling out to `notify-send'."
  (call-process-shell-command
   (format "notify-send -t 2000 \"%s\" \"%s\"" title message)))

(use-package treemacs)
;;  (treemacs-enable)

(tool-bar-mode 0)
(menu-bar-mode 0)
(scroll-bar-mode -1)
(global-linum-mode t)
(column-number-mode t)
(add-hook 'before-save-hook 'delete-trailing-whitespace)

(setq frame-title-format "%b")
;; (setq frame-title-format '((:eval (projectile-project-name))))

(global-prettify-symbols-mode t)

(use-package abyss-theme)
;; (load-theme 'abyss t)

(use-package moody
  :config
  (setq x-underline-at-descent-line t
	moody-mode-line-height 30))

(setq zb/default-font "Source Code Pro")
(setq zb/default-font-size 12)
(setq zb/current-font-size zb/default-font-size)

(setq zb/font-change-increment 1.1)

(defun zb/font-code ()
  "Return a string representing the current font (like \"Source Code Pro-14\")."
  (concat zb/default-font "-" (number-to-string zb/current-font-size)))

(defun zb/set-font-size ()
  "Set the font to `zb/default-font' at `zb/current-font-size'.
Set that for the current frame, and also make it the default for
other, future frames."
  (let ((font-code (zb/font-code)))
    (if (assoc 'font default-frame-alist)
	(setcdr (assoc 'font default-frame-alist) font-code)
(add-to-list 'default-frame-alist (cons 'font font-code)))
    (set-frame-font font-code)))

(defun zb/reset-font-size ()
  "Change font size back to `zb/default-font-size'."
  (interactive)
  (setq zb/current-font-size zb/default-font-size)
  (zb/set-font-size))

(defun zb/increase-font-size ()
  "Increase current font size by a factor of `zb/font-change-increment'."
  (interactive)
  (setq zb/current-font-size
	(ceiling (* zb/current-font-size zb/font-change-increment)))
  (zb/set-font-size))

(defun zb/decrease-font-size ()
  "Decrease current font size by a factor of `zb/font-change-increment', down to a minimum size of 1."
  (interactive)
  (setq zb/current-font-size
	(max 1
	(floor (/ zb/current-font-size zb/font-change-increment))))
  (zb/set-font-size))

(define-key global-map (kbd "C-)") 'zb/reset-font-size)
(define-key global-map (kbd "C-+") 'zb/increase-font-size)
(define-key global-map (kbd "C--") 'zb/decrease-font-size)
(zb/reset-font-size)

(global-hl-line-mode)

(use-package diff-hl
  :config
  (add-hook 'prog-mode-hook 'turn-on-diff-hl-mode)
  (add-hook 'vc-dir-mode-hook 'turn-on-diff-hl-mode))

(use-package company)
(add-hook 'after-init-hook 'global-company-mode)

(global-set-key (kbd "M-/") 'company-complete-common)

(use-package let-alist)
(use-package flycheck
  :hook ruby-mode)

(use-package magit
  :bind
  ("C-x g" . magit-status)

  :config
  (use-package with-editor)

  (setq magit-push-always-verify nil
	git-commit-summary-max-length 50))

(use-package git-timemachine)

(use-package projectile
  :bind
  ("C-c v" . projectile-ag)

  :config
  (define-key projectile-mode-map (kbd "C-c p") 'projectile-command-map)

  (setq projectile-completion-system 'ivy
	projectile-switch-project-action 'projectile-dired
	projectile-require-project-root nil))

(setq-default tab-width 2)

(use-package subword
  :config (global-subword-mode 1))

(setq compilation-scroll-output t)

(use-package eglot)

(use-package css-mode
  :config
  (setq css-indent-offset 2))

(use-package scss-mode
  :config
  (setq scss-compile-at-save nil))

(use-package less-css-mode)

(use-package go-mode)
(use-package go-errcheck)
(use-package company-go)

(setenv "GOPATH" "/home/zach/code/go")
(zb/append-to-path (concat (getenv "GOPATH") "/bin"))

(setq gofmt-command "goimports")
(add-hook 'before-save-hook 'gofmt-before-save)

(add-hook 'go-mode-hook
          (lambda ()
            (set (make-local-variable 'company-backends)
                 '(company-go))
            (company-mode)
            (if (not (string-match "go" compile-command))
                (set (make-local-variable 'compile-command)
                     "go build -v && go test -v && go vet"))
            (flycheck-mode)))

(use-package haskell-mode)

(add-hook 'haskell-mode-hook
          (lambda ()
            (haskell-doc-mode)
            (turn-on-haskell-indent)))

(zb/append-to-path "~/.cabal/bin")

(use-package coffee-mode)

(setq js-indent-level 2)

(add-hook 'coffee-mode-hook
          (lambda ()
            (yas-minor-mode 1)
            (setq coffee-tab-width 2)))

(use-package paredit)

(use-package rainbow-delimiters)

(setq lispy-mode-hooks
      '(clojure-mode-hook
        emacs-lisp-mode-hook
        lisp-mode-hook
        scheme-mode-hook))

(dolist (hook lispy-mode-hooks)
  (add-hook hook (lambda ()
                   (setq show-paren-style 'expression)
                   (paredit-mode)
                   (rainbow-delimiters-mode))))

(use-package eldoc
  :config
  (add-hook 'emacs-lisp-mode-hook 'eldoc-mode))

(use-package flycheck-package)

(eval-after-load 'flycheck
  '(flycheck-package-setup))

(use-package python-mode)

(zb/append-to-path "~/.local/bin")

(use-package elpy)
(elpy-enable)

(add-hook 'elpy-mode-hook 'flycheck-mode)

(use-package py-autopep8)
(require 'py-autopep8)
(add-hook 'elpy-mode-hook 'py-autopep8-enable-on-save)

(use-package company-jedi)
(add-to-list 'company-backends 'company-jedi)

(add-hook 'python-mode-hook 'jedi:setup)
(setq jedi:complete-on-dot t)

(use-package rust-mode
  :config
  (zb/append-to-path "~/.cargo/bin")
  (setq rust-format-on-save t))

(add-hook 'sh-mode-hook
          (lambda ()
            (setq sh-basic-offset 2
                  sh-indentation 2)))

(use-package yaml-mode)

(defun zb/term-paste (&optional string)
  (interactive)
  (process-send-string
   (get-buffer-process (current-buffer))
   (if string string (current-kill 0))))

(add-hook 'term-mode-hook
          (lambda ()
            (goto-address-mode)
            (define-key term-raw-map (kbd "C-y") 'zb/term-paste)
            (define-key term-raw-map (kbd "<mouse-2>") 'zb/term-paste)
            (define-key term-raw-map (kbd "M-o") 'other-window)
            (setq yas-dont-activate t)))

(setq initial-major-mode 'org-mode)

(use-package org-bullets
  :init
  (add-hook 'org-mode-hook 'org-bullets-mode))

(setq org-ellipsis "⤵")

(setq org-src-fontify-natively t)

(setq org-src-tab-acts-natively t)

(setq org-src-window-setup 'current-window)

(add-to-list 'org-structure-template-alist
             '("el" . "src emacs-lisp"))

(setq org-adapt-indentation nil)

(setq org-directory "~/Documents/org")

(defun org-file-path (filename)
  "Return the absolute address of an org file, given its relative name."
  (concat (file-name-as-directory org-directory) filename))

(setq org-inbox-file "~/sync/Dropbox/inbox.org")
(setq org-index-file (org-file-path "index.org"))
(setq org-archive-location
      (concat (org-file-path "archive.org") "::* From %s"))

(defun zb/copy-tasks-from-inbox ()
  (when (file-exists-p org-inbox-file)
    (save-excursion
      (find-file org-index-file)
      (goto-char (point-max))
      (insert-file-contents org-inbox-file)
      (delete-file org-inbox-file))))

(setq org-agenda-files (list org-index-file
                             (org-file-path "events.org")
                             (org-file-path "goals.org")
                             (org-file-path "recurring-events.org")
                             (org-file-path "work.org")))

(defun zb/mark-done-and-archive ()
  "Mark the state of an org-mode item as DONE and archive it."
  (interactive)
  (org-todo 'done)
  (org-archive-subtree))

(define-key org-mode-map (kbd "C-c C-x C-s") 'zb/mark-done-and-archive)

(setq org-log-done 'time)

(setq org-enforce-todo-dependencies t)
(setq org-enforce-todo-checkbox-dependencies t)

(setq org-agenda-start-on-weekday nil)

(require 'org-habit)

(defun zb/org-skip-subtree-if-priority (priority)
  "Skip an agenda subtree if it has a priority of PRIORITY.

PRIORITY may be one of the characters ?A, ?B, or ?C."
  (let ((subtree-end (save-excursion (org-end-of-subtree t)))
        (pri-value (* 1000 (- org-lowest-priority priority)))
        (pri-current (org-get-priority (thing-at-point 'line t))))
    (if (= pri-value pri-current)
        subtree-end
      nil)))

(defun zb/org-skip-subtree-if-habit ()
  "Skip an agenda entry if it has a STYLE property equal to \"habit\"."
  (let ((subtree-end (save-excursion (org-end-of-subtree t))))
    (if (string= (org-entry-get nil "STYLE") "habit")
        subtree-end
      nil)))

(setq org-agenda-custom-commands
      '(("p" "Personal agenda"
         ((agenda "")
          (todo "TODO"
                ((org-agenda-skip-function '(or (zb/org-skip-subtree-if-priority ?A)
                                                (zb/org-skip-subtree-if-habit)))
                 (org-agenda-overriding-header "Other tasks:")))
          (todo "PENDING"
                ((org-agenda-skip-function '(zb/org-skip-subtree-if-priority ?A))
                 (org-agenda-overriding-header "Pending:")))
          (todo "BLOCKED"
                ((org-agenda-skip-function '(zb/org-skip-subtree-if-priority ?A))
                 (org-agenda-overriding-header "Blocked:")))))))

(defun zb/dashboard ()
  (interactive)
  (zb/copy-tasks-from-inbox)
  (find-file org-index-file)
  (org-agenda nil "p"))

(global-set-key (kbd "C-c d") 'zb/dashboard)

(setq org-capture-templates
      '(("b" "Blog idea"
         entry
         (file "~/Documents/notes/blog-ideas.org")
         "* %?\n")

        ("c" "Contact"
         entry
         (file "~/Documents/contacts.org")
         "* %(org-contacts-template-name)
:PROPERTIES:
:ADDRESS: %^{123 Fake St., City, ST 12345}
:PHONE: %^{555-555-5555}
:EMAIL: %(org-contacts-template-email)
:NOTE: %^{note}
:END:")

        ("d" "Delivery" entry
         (file+headline "~/Documents/org/events.org" "Deliveries")
         "** %?\n   SCHEDULED: %t\n")

        ("e" "Email" entry
         (file+headline org-index-file "Inbox")
         "* TODO %?\n\n%a\n\n")

        ("f" "Finished book"
         table-line (file "~/Documents/notes/books-read.org")
         "| %^{Title} | %^{Author} | %u |")

        ("s" "Subscribe to an RSS feed"
         plain
         (file "~/Documents/rss-feeds.org")
         "*** [[%^{Feed URL}][%^{Feed name}]]")

        ("t" "Todo"
         entry
         (file+headline org-index-file "Inbox")
         "* TODO %?\n:PROPERTIES:\nCREATED: %u\n:END:\n")))

(setq org-refile-use-outline-path t)
(setq org-outline-path-complete-in-steps nil)

(define-key global-map "\C-cl" 'org-store-link)
(define-key global-map "\C-ca" 'org-agenda)
(define-key global-map "\C-cc" 'org-capture)

(defun zb/open-index-file ()
  "Open the master org TODO list."
  (interactive)
  (zb/copy-tasks-from-inbox)
  (find-file org-index-file)
  (flycheck-mode -1)
  (end-of-buffer))

(global-set-key (kbd "C-c i") 'zb/open-index-file)

(defun org-capture-todo ()
  (interactive)
  (org-capture :keys "t"))

(global-set-key (kbd "M-n") 'org-capture-todo)
(add-hook 'gfm-mode-hook
          (lambda () (local-set-key (kbd "M-n") 'org-capture-todo)))
(add-hook 'haskell-mode-hook
          (lambda () (local-set-key (kbd "M-n") 'org-capture-todo)))

(defun zb/open-work-file ()
  "Open the work TODO list."
  (interactive)
  (find-file (org-file-path "work.org"))
  (flycheck-mode -1)
  (end-of-buffer))

(global-set-key (kbd "C-c w") 'zb/open-work-file)

(require 'ox-md)
(require 'ox-beamer)

(setq org-confirm-babel-evaluate nil)

(use-package htmlize)

(setq org-export-with-smart-quotes t)

(setq org-html-postamble nil)

(setq browse-url-browser-function 'browse-url-generic
      browse-url-generic-program "firefox")

(setenv "BROWSER" "firefox")

(setq org-latex-pdf-process
      '("xelatex -shell-escape -interaction nonstopmode -output-directory %o %f"
        "xelatex -shell-escape -interaction nonstopmode -output-directory %o %f"
        "xelatex -shell-escape -interaction nonstopmode -output-directory %o %f"))

(add-to-list 'org-latex-packages-alist '("" "minted"))
(setq org-latex-listings 'minted)

(setq TeX-PDF-mode t)

(setq-default TeX-engine 'default)
(setq TeX-parse-self t)
;; (latex-preview-pane-enable)

(use-package auctex-lua)
(use-package auctex-latexmk)
(use-package textmate)
(use-package cdlatex)
(use-package latex-extra)
(use-package latex-math-preview)
(use-package latex-unicode-math-mode)
(use-package ebib)

(add-hook 'LaTeX-mode-hook
          (lambda ()
            (LaTeX-math-mode)
            (setq TeX-master t)))

(defun zb/visit-emacs-config ()
  (interactive)
  (find-file "~/.emacs.d/configuration.org"))

(global-set-key (kbd "C-c e") 'zb/visit-emacs-config)

(defun zb/kill-current-buffer ()
  "Kill the current buffer without prompting."
  (interactive)
  (kill-buffer (current-buffer)))

(global-set-key (kbd "C-x k") 'zb/kill-current-buffer)

(zb/append-to-path "/usr/local/bin")

(save-place-mode t)

(setq-default indent-tabs-mode nil)

(use-package which-key
  :config (which-key-mode))

(use-package yasnippet)

(setq yas-snippet-dirs '("~/.emacs.d/snippets/text-mode"))
(yas-global-mode 1)

(setq yas-indent-line 'auto)

(defun zb/split-window-below-and-switch ()
  "Split the window horizontally, then switch to the new pane."
  (interactive)
  (split-window-below)
  (balance-windows)
  (other-window 1))

(defun zb/split-window-right-and-switch ()
  "Split the window vertically, then switch to the new pane."
  (interactive)
  (split-window-right)
  (balance-windows)
  (other-window 1))

(global-set-key (kbd "C-x 2") 'zb/split-window-below-and-switch)
(global-set-key (kbd "C-x 3") 'zb/split-window-right-and-switch)

(projectile-global-mode)

(use-package engine-mode)
(require 'engine-mode)

(defengine duckduckgo
  "https://duckduckgo.com/?q=%s"
  :keybinding "d")

(defengine github
  "https://github.com/search?ref=simplesearch&q=%s"
  :keybinding "g")

(defengine google
  "http://www.google.com/search?ie=utf-8&oe=utf-8&q=%s")

(defengine rfcs
  "http://pretty-rfc.herokuapp.com/search?q=%s")

(defengine stack-overflow
  "https://stackoverflow.com/search?q=%s"
  :keybinding "s")

(defengine wikipedia
  "http://www.wikipedia.org/search-redirect.php?language=en&go=Go&search=%s"
  :keybinding "w")

(defengine wiktionary
  "https://www.wikipedia.org/search-redirect.php?family=wiktionary&language=en&go=Go&search=%s")

(defengine youtube
  "https://www.youtube.com/results?search_query=%s")

(engine-mode t)

(define-key input-decode-map "\e[1;2A" [S-up])
